### PR TITLE
Initialise Nusense HardwareIO

### PR DESCRIPTION
**Pull request by jpptm**
Things the PR does go here

- Adds necessary HardwareIO code for NUsense, our very own subcontroller for the robots

### Pre-PR Checklist:

Have you
- [x] Updated relevant module READMEs
- [x] Added a descriptive title and relevant labels to the PR

### Things left to do

- [x] Test functionalities as we go so we don't go insane debugging later on
- [x] Replace c code with c++

### Required functionalities
- [x] Serialising protobuf to nbs msg and vice versa
- [x] Spam msgs 90 Hz
- [x] Stream reactor

### CODE REVIEW

File: module/platform/NUsense/HardwareIO/data/config/nugus1/HardwareIO.yaml [hunk 1]
```diff
@@ -0,0 +1,92 @@
...
+
+bus_reset_wait_time_us: 0
+byte_wait: 1200
+packet_wait: 100000
+
+servos:
+  - # [0]  R_SHOULDER_PITCH
+    direction: -1
+    offset: 0.0
+    simulated: false
```
Comment (TrentHouliston) [inline module/platform/NUsense/HardwareIO/data/config/nugus1/HardwareIO.yaml:14]: This should all be done on the NUsense I think

File: module/platform/NUsense/HardwareIO/data/config/nugus1/HardwareIO.yaml [hunk 2]
```diff
@@ -0,0 +1,92 @@
+battery:
+  charged_voltage: 16.8
+  nominal_voltage: 14.8
+  flat_voltage: 14.0
+
+bus_reset_wait_time_us: 0
+byte_wait: 1200
+packet_wait: 100000
```
Comment (TrentHouliston) [inline module/platform/NUsense/HardwareIO/data/config/nugus1/HardwareIO.yaml:8]: These don't make sense for this

File: module/platform/NUsense/HardwareIO/data/config/nugus1/HardwareIO.yaml [hunk 3]
```diff
@@ -0,0 +1,92 @@
+battery:
+  charged_voltage: 16.8
+  nominal_voltage: 14.8
+  flat_voltage: 14.0
```
Comment (TrentHouliston) [inline module/platform/NUsense/HardwareIO/data/config/nugus1/HardwareIO.yaml:4]: NUsense itself will handle all this

File: module/platform/NUsense/HardwareIO/src/NUsense/SIProcessor.cpp [hunk 4]
```diff
@@ -0,0 +1,59 @@
...
+        packet_data.push_back(0xE2);
+        packet_data.push_back(0x98);
+        packet_data.push_back(0xA2);
+
+        // Serialise protobuf message to string then add everything to our vector
+        std::string protobuf_bytes = msg.SerializeToString();
+
+        for (const auto& byte : protobuf_bytes) {
+            packet_data.push_back(static_cast<uint8_t>(byte));
+        }
```
Comment (TrentHouliston) [inline module/platform/NUsense/HardwareIO/src/NUsense/SIProcessor.cpp:22]: Use `vector.insert` function instead of doing a loop

File: module/platform/NUsense/HardwareIO/src/NUsense/SIProcessor.cpp [hunk 5]
```diff
@@ -0,0 +1,59 @@
...
+        // We insert everything to a string stream
+        std::vector<uint8_t> packets;
+
+        // Add header
+        packet_data.push_back(0xE2);
+        packet_data.push_back(0x98);
+        packet_data.push_back(0xA2);
+
+        // Serialise protobuf message to string then add everything to our vector
+        std::string protobuf_bytes = msg.SerializeToString();
```
Comment (TrentHouliston) [inline module/platform/NUsense/HardwareIO/src/NUsense/SIProcessor.cpp:18]: This is going to be a Neutron not a protobuf, so you need to use the NUClear serialisation method

File: shared/message/platform/RawSensorsV2.proto [hunk 6]
```diff
@@ -0,0 +1,194 @@
...
+ message ButtonLeftDown {}
+
+ message ButtonLeftUp {}
+
+ message ButtonMiddleDown {}
+
+ message ButtonMiddleUp {}
+
+ // Reset sensors and servo targets in Webots
+ message ResetWebotsServos {}
```
Comment (TrentHouliston) [inline shared/message/platform/RawSensorsV2.proto:194]: I feel like these belong elsewhere

File: module/platform/NUsense/HardwareIO/src/NUsense/SIProcessor.cpp (hunk 5 above)
Comment (jpptm) [reply module/platform/NUsense/HardwareIO/src/NUsense/SIProcessor.cpp:18]: Does the nuclear one agree with the protobuf one? I just wanna make sure as nanopb agrees with vanilla protobuf only

File: shared/message/platform/RawSensorsV2.proto (hunk 6 above)
Comment (jpptm) [reply shared/message/platform/RawSensorsV2.proto:194]: They definitely do 👌 Thank you!

**Comment by jpptm** [comment]
TODO open /dev/ttyACM0 or USB0 (config val) using the NUClear uart utility for it and send dummy messages for testing

File: docker/Dockerfile [hunk 7]
```diff
@@ -358,8 +358,10 @@ RUN chmod 440 /etc/sudoers.d/user
 # Create the user, and setup sudo so no password is required
 ARG user_uid=1000
 ARG user_gid=$user_uid
-RUN groupadd --gid $user_gid nubots \
-    && useradd --uid $user_uid --gid $user_gid -m nubots
+ARG dialout_gid=20
+RUN groupadd --gid $dialout_gid dialout \
+    && groupadd --gid $user_gid nubots \
+    && useradd --uid $user_uid --gid nubots -G dialout -m nubots
```
Comment (Bidski) [inline docker/Dockerfile:364]:
```suggestion
ARG dialout_gid=20

# useradd -G is not sufficient to add the nubots user to the dialout 
# group as docker manages to lose supplementary groups when changing 
# to the nubots user.
# Also need to add "--group-add dialout" to the docker run command in run.py
RUN groupadd --gid $dialout_gid dialout \
    && groupadd --gid $user_gid nubots \
    && useradd --uid $user_uid --gid nubots -G dialout -m nubots
```

File: tools/utility/dockerise/run.py [hunk 8]
```diff
@@ -144,6 +144,8 @@ def _run(**kwargs):
             "--env",
             f"EDITOR={os.environ.get('EDITOR', 'nano')}",
             "--privileged",
+            "--group-add",
+            "dialout",
```
Comment (Bidski) [inline tools/utility/dockerise/run.py:148]:
```suggestion
            # Also need to create the group in the Dockerfile
            "--group-add",
            "dialout",
```

File: module/platform/NUsense/HardwareIO/src/HardwareIO.cpp [hunk 9]
```diff
@@ -0,0 +1,55 @@
...
+        });
+
+        on<Startup>().then("NUSense HardwareIO Startup", [this] {
+            // Make sure the port is closed before we open it
+            if (nusense.connected()){
+                nusense.close();
+            }
+
+            // If /dev/ttyACM0 is acting up, follow the link below
+            // https://stackoverflow.com/questions/40951728/avrdude-ser-open-cant-open-device-dev-ttyacm0-device-or-resource-busy
```
Comment (Bidski) [inline module/platform/NUsense/HardwareIO/src/HardwareIO.cpp:27]:
```suggestion
```

File: module/platform/NUsense/HardwareIO/src/HardwareIO.cpp [hunk 10]
```diff
@@ -0,0 +1,55 @@
...
+            nusense.open(cfg.nusense.port, cfg.nusense.baud);
```
Comment (Bidski) [inline module/platform/NUsense/HardwareIO/src/HardwareIO.cpp:28]: I would issue caution when setting the baud rate, I don't think the NUSense is using a baud rate and I am not sure what would happen if you try to set one?

**Review by Bidski — CHANGES_REQUESTED** [review]

File: module/platform/NUsense/HardwareIO/src/HardwareIO.cpp (hunk 9 above)
Comment (jpptm) [reply module/platform/NUsense/HardwareIO/src/HardwareIO.cpp:27]: I'm not sure whether we should remove it. I put the link there because it was a fix for nix based machines not seeing the port when we connect the USB to it

File: tools/utility/dockerise/run.py (hunk 8 above)
Comment (TrentHouliston) [reply tools/utility/dockerise/run.py:174]: Also, you can't do this, it makes any computer (i.e. mine) no longer able to build the code at all.

**Review by TrentHouliston — CHANGES_REQUESTED** [review]

File: tools/utility/dockerise/run.py (hunk 8 above)
Comment (jpptm) [reply tools/utility/dockerise/run.py:174]: @TrentHouliston  May I ask why? @Claegtun has been testing and building with this branch and it worked.

Comment (TrentHouliston) [reply tools/utility/dockerise/run.py:174]: I'm gonna guess he doesn't use a Mac

Comment (jpptm) [reply tools/utility/dockerise/run.py:174]: He uses pop afaik, Im not really sure what I should do instead of this so it builds with anything. Is there a dialout equivalent for mac?

Comment (Claegtun) [reply tools/utility/dockerise/run.py:174]: > He uses pop afaik, Im not really sure what I should do instead of this so it builds with anything. Is there a dialout equivalent for mac?

I actually use Mint.

> @TrentHouliston May I ask why? @Claegtun has been testing and building in this branch with these lines and it worked.

Also, I don't think that I have been building from this branch as is. I had to merge with main to get it to build properly. I am not sure if that is to what @TrentHouliston is referring.

Comment (Bidski) [reply tools/utility/dockerise/run.py:174]: How does this break it on Mac? Is it just that dialout is the wrong group?

Comment (TrentHouliston) [reply tools/utility/dockerise/run.py:174]: Dialout group doesn't exist so it just fails

Comment (Bidski) [reply tools/utility/dockerise/run.py:174]: According to this you shouldn't need it on the host?
https://github.com/docker/cli/issues/4714#issuecomment-1881571231

Comment (Bidski) [reply tools/utility/dockerise/run.py:174]: Have you rebuilt the image?

Comment (Bidski) [reply tools/utility/dockerise/run.py:174]: Although there looks like there is a change to platform.py that is missing (but this missing should probably work in your favour)

File: module/platform/NUsense/HardwareIO/src/HardwareIO.cpp (hunk 10 above)
Comment (JohanneMontano) [reply module/platform/NUsense/HardwareIO/src/HardwareIO.cpp]: NUSense just apparently ignores the baud rate set by this.

File: module/platform/NUsense/HardwareIO/src/util/packet_handler.cpp
Comment (Claegtun) [inline module/platform/NUsense/HardwareIO/src/util/packet_handler.cpp:1]: Change the file's name to Pascal case, i.e. PacketHandler.cpp.

File: module/platform/NUsense/HardwareIO/src/util/packet_handler.hpp
Comment (Claegtun) [inline module/platform/NUsense/HardwareIO/src/util/packet_handler.hpp:1]: Change the file's name to Pascal case, i.e. PacketHandler.hpp.

**Review by Claegtun — COMMENTED** [review]
Should remove all the debug-logs and clean this up so that it is ready to be merged or to merge with Dex's branch. 

Also, the PacketHandler code is very rough and ready given that it is code copied from NUSense. It will have to change quite a bit before this will get merged with main, e.g. the `memcpy`'s were a C-like compromise because of ST's function being written in C. But given that this is entirely C++, it should be `std::copy`. 

Also, I am not sure whether a ring-buffer is needed. We used the ring-buffer in the NUSense since the callback needed to be short, and the decoding was done elsewhere. However, given that both the receiving and the decoding are done in the same scope, i.e. `on<IO>`, it seems very redundant. So, we may just have one function that takes `nusense_usb_bytes`.

However, for the sake of getting this merged with Dex's branch and then working rough and ready only within this branch, I may allow the ring-buffer, the `memcpy`'s, etc. since it is working, and I want us to push forward.

So, watch this space before merging with main.

**Comment by JohanneMontano** [comment]
NOTE: This must go in main together with pr #10 in NUcontroller, otherwise it wont work

**Comment by JohanneMontano** [comment]
People should now be able to  `git pull` safely

File: module/platform/NUSense/HardwareIO/src/HardwareIO.cpp [hunk 11]
```diff
@@ -0,0 +1,222 @@
...
+                    log<NUClear::DEBUG>(fmt::format("    pvelo: {}", val.present_velocity));
+                    log<NUClear::DEBUG>(fmt::format("     ppos: {}", val.present_position));
+                    log<NUClear::DEBUG>(fmt::format("     gpwm: {}", val.goal_pwm));
+                    log<NUClear::DEBUG>(fmt::format("    gcurr: {}", val.goal_current));
+                    log<NUClear::DEBUG>(fmt::format("    gvelo: {}", val.goal_velocity));
+                    log<NUClear::DEBUG>(fmt::format("     gpos: {}", val.goal_position));
+                    log<NUClear::DEBUG>(fmt::format("  voltage: {}", val.voltage));
+                    log<NUClear::DEBUG>(fmt::format("     temp: {}", val.temperature));
+                }
+            }
```
Comment (ysims) [inline module/platform/NUSense/HardwareIO/src/HardwareIO.cpp:186]: theres no need to wrap this in an if

File: shared/utility/reactor/StreamReactor.hpp [hunk 12]
```diff
@@ -427,9 +436,15 @@ namespace utility::reactor {
          * passes it to the parser. Once the parser finds a packet it emits that packet.
          */
         void process(const int& fd) {
+            std::stringstream debug_string;
```
Comment (ysims) [inline shared/utility/reactor/StreamReactor.hpp:439]:
```suggestion
            std::stringstream debug_string{};
```

File: module/platform/NUSense/HardwareIO/src/NUSenseParser.hpp [hunk 13]
```diff
@@ -0,0 +1,40 @@
...
+#include <vector>
+
+namespace module::platform::NUSense {
+
+    struct NUSenseFrame {
+        std::array<uint8_t, 3> header;  //< NBS Header, should be 0xE2, 0x98, 0xA2
+        uint32_t size;                  //< Size of the payload after this field
+        uint64_t timestamp;             //< Timestamp of the message
+        uint64_t hash;                  //< Hash of the message type
+        std::vector<uint8_t> payload;   //< The payload of the message
```
Comment (ysims) [inline module/platform/NUSense/HardwareIO/src/NUSenseParser.hpp:16]: initialise?

File: shared/utility/strutil/strutil.hpp [hunk 14]
```diff
@@ -1,125 +1,280 @@
...
- *
- * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
- * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
- * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
- * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
- * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
- * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
- * SOFTWARE.
- */
-
```
Comment (ysims) [inline shared/utility/strutil/strutil.hpp:27]: add licenses

**Review by ysims — COMMENTED** [review]
Check that you have the latest HardwareIOs from OpenCR for the robots. There's been some changes recently.

File: shared/utility/reactor/StreamReactor.hpp [hunk 15]
```diff
@@ -443,6 +458,9 @@ namespace utility::reactor {
                     log<NUClear::WARN>(ex.what());
                 }
             }
+            if (log_level <= NUClear::TRACE) {
+                log<NUClear::TRACE>("Read:", debug_string.str());
+            }
```
[the remark this answers is not in the document]
Comment (TrentHouliston) [reply shared/utility/reactor/StreamReactor.hpp:463]: This one is here on purpose, should probably have a comment.
Since this will be dumping so much data to log you really don't want to be doing all the work involved in making a log line for every single read/write you get.

File: module/platform/NUSense/HardwareIO/src/HardwareIO.cpp (hunk 11 above)
Comment (TrentHouliston) [reply module/platform/NUSense/HardwareIO/src/HardwareIO.cpp:186]: As a counter point to my other comment with the trace line, this one doesn't need the if statement

File: module/platform/NUSense/HardwareIO/src/HardwareIO.hpp [hunk 16]
```diff
@@ -59,7 +59,10 @@ namespace module::platform::NUSense {
         NUgus nugus{};
 
     private:
-        /// @brief Send message to StreamReactor so it can write the data to NUSense using uart
+        /// @brief Send a TransmitData message containing an nbs packet to StreamReactor so it can write the data to
+        /// NUSense.
+        /// @tparam packet A const reference to a protobuf message of type T.
+        ///                This gets serialised and turned to a vector of bytes before sending to NUSense.
```
Comment (ysims) [inline module/platform/NUSense/HardwareIO/src/HardwareIO.hpp:65]: tparam is the template parameter (`@tparam T`), different to normal param (`@param packet`)

File: module/platform/NUSense/HardwareIO/src/HardwareIO.hpp [hunk 17]
```diff
@@ -0,0 +1,104 @@
...
+        NUgus nugus{};
+
+    private:
+        /// @brief Send a TransmitData message containing an nbs packet to StreamReactor so it can write the data to
+        /// NUSense.
+        /// @tparam packet A const reference to a protobuf message of type T.
+        ///                This gets serialised and turned to a vector of bytes before sending to NUSense.
+        template <typename T>
+        void send_packet(const T& packet) {
+            // Serialize the packet
```
Comment (ysims) [inline module/platform/NUSense/HardwareIO/src/HardwareIO.hpp:68]:
```suggestion
            // Serialise the packet
```

File: module/platform/NUSense/HardwareIO/src/NUSenseParser.hpp [hunk 18]
```diff
@@ -0,0 +1,40 @@
...
+    };
+
+    struct NUSenseParser {
+        /**
+         * Parses bytes from NUSense one byte at a time and returns a NUSenseFrame when a complete frame is read
+         *
+         * @param byte The next byte to parse
+         *
+         * @return A NUSenseFrame if a complete frame has been read, otherwise nullptr
+         */
```
Comment (ysims) [inline module/platform/NUSense/HardwareIO/src/NUSenseParser.hpp:26]:
```suggestion
         /// @brief Parses bytes from NUSense one byte at a time and returns a NUSenseFrame when a complete frame is read
         /// @param byte The next byte to parse
         /// @return A NUSenseFrame if a complete frame has been read, otherwise nullptr
```

File: module/platform/NUSense/HardwareIO/src/HardwareIO.hpp (hunk 16 above)
Comment (JohanneMontano) [reply module/platform/NUSense/HardwareIO/src/HardwareIO.hpp]: aha gotcha thank you thats makes more sense

File: shared/utility/strutil/strutil.hpp [hunk 19]
```diff
@@ -29,97 +29,307 @@
...
+     * @brief Checks if a string ends with a specific substring.
+     *
+     * This function compares the given string `str` with the specified `ending` substring
+     * and determines whether `str` ends with `ending`. The comparison is case-sensitive.
+     *
+     * @param str The string to check for the ending substring.
+     * @param ending The substring to search for at the end of the string.
+     * @return `true` if `str` ends with `ending`, `false` otherwise.
+     */
     // http://stackoverflow.com/a/874160/1387006
```
Comment (ysims) [inline shared/utility/strutil/strutil.hpp:57]:
```suggestion
    /**
     * @brief Checks if a string ends with a specific substring.
     *
     * This function compares the given string `str` with the specified `ending` substring
     * and determines whether `str` ends with `ending`. The comparison is case-sensitive.
     * http://stackoverflow.com/a/874160/1387006
     *
     * @param str The string to check for the ending substring.
     * @param ending The substring to search for at the end of the string.
     * @return `true` if `str` ends with `ending`, `false` otherwise.
     */
```

File: shared/utility/strutil/strutil.hpp [hunk 20]
```diff
@@ -29,97 +29,307 @@
...
+     * @brief The overloaded version of ends_with which takes vectors of strings instead of single strings.
+     *
+     * This function compares the given string `str` with the specified `ending` substring in the vector endings
+     * and determines whether `str` ends with `ending`. The comparison is case-sensitive.
+     *
+     * @param str The string to check for the ending substring.
+     * @param ending The substring to search for at the end of the string.
+     * @return `true` if `str` ends with `ending`, `false` otherwise.
+     */
+    // http://stackoverflow.com/a/874160/1387006
```
Comment (ysims) [inline shared/utility/strutil/strutil.hpp:73]: this should also be in the main function description

**Review by ysims — APPROVED** [review]

File: module/input/FakeCamera/src/FakeCamera.cpp [hunk 21]
```diff
@@ -82,7 +82,7 @@ namespace module::input {
 
                     // Traverse specified directory and locate all image/lens pairs
                     for (const auto& p : std::filesystem::recursive_directory_iterator(config.image_folder)) {
-                        if (utility::strutil::endsWith(p.path(), std::vector<std::string>{".jpg", ".jpeg"})) {
+                        if (utility::strutil::ends_with(p.path(), std::vector<std::string>{".jpg", ".jpeg"})) {
```
Comment (Bidski) [inline module/input/FakeCamera/src/FakeCamera.cpp:85]: Seems unrelated to this PR

File: module/platform/NUSense/HardwareIO/data/config/nugus1/HardwareIO.yaml
Comment (Bidski) [inline module/platform/NUSense/HardwareIO/data/config/nugus1/HardwareIO.yaml:1]: Remove things from this file that are identical to the parent file, same with the other robot-specific configs

File: module/platform/NUSense/HardwareIO/src/NUSenseParser.cpp [hunk 22]
```diff
@@ -0,0 +1,64 @@
+#include "NUSenseParser.hpp"
+
+namespace module::platform::NUSense {
+
+    uint32_t read_le_32(const uint8_t* ptr) {
```
Comment (Bidski) [inline module/platform/NUSense/HardwareIO/src/NUSenseParser.cpp:5]:
```suggestion
    /**
     * @brief .....
     */
    uint32_t read_le_32(const uint8_t* ptr) {
```

File: module/platform/NUSense/HardwareIO/src/NUSenseParser.cpp [hunk 23]
```diff
@@ -0,0 +1,64 @@
...
+        return (uint32_t(ptr[0]) << 0) | (uint32_t(ptr[1]) << 8) | (uint32_t(ptr[2]) << 16) | (uint32_t(ptr[3]) << 24);
+    }
+
+    uint64_t read_le_64(const uint8_t* ptr) {
```
Comment (Bidski) [inline module/platform/NUSense/HardwareIO/src/NUSenseParser.cpp:9]:
```suggestion
    /**
     * @brief .....
     */
    uint64_t read_le_64(const uint8_t* ptr) {
```

File: module/platform/NUSense/HardwareIO/src/NUSenseParser.hpp [hunk 24]
```diff
@@ -0,0 +1,37 @@
...
+
+#include <array>
+#include <cstdint>
+#include <memory>
+#include <vector>
+
+namespace module::platform::NUSense {
+
+    struct NUSenseFrame {
+        std::array<uint8_t, 3> header{};  //< NBS Header, should be 0xE2, 0x98, 0xA2
```
Comment (Bidski) [inline module/platform/NUSense/HardwareIO/src/NUSenseParser.hpp:12]: Same for the other fields
```suggestion
        /// NBS Header, should be 0xE2, 0x98, 0xA2
        std::array<uint8_t, 3> header{};
```

File: module/platform/NUSense/HardwareIO/src/NUgus.cpp [hunk 25]
```diff
@@ -0,0 +1,33 @@
...
+ * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+ * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+ * SOFTWARE.
+ */
+#include "NUgus.hpp"
+
+namespace module::platform::NUSense {
+
+    NUgus::NUgus() {}
```
Comment (Bidski) [inline module/platform/NUSense/HardwareIO/src/NUgus.cpp:31]: Not a lot of point in having this here? Not much point in having a default constructor at all if it is just going to be empty

File: module/tools/FirmwareInstaller/src/FirmwareInstaller.cpp [hunk 26]
```diff
@@ -56,8 +56,8 @@ namespace module::tools {
 
             for (const auto& f : config["firmwares"].config) {
                 std::pair<std::string, std::string> name;
-                name.first  = utility::strutil::toUpper(f["device"].as<std::string>());
-                name.second = utility::strutil::toUpper(f["battery"].as<std::string>());
+                name.first  = utility::strutil::to_upper(f["device"].as<std::string>());
+                name.second = utility::strutil::to_pper(f["battery"].as<std::string>());
```
Comment (Bidski) [inline module/tools/FirmwareInstaller/src/FirmwareInstaller.cpp:60]: Seems unrelated to this PR?

**Review by Bidski — CHANGES_REQUESTED** [review]

File: module/platform/NUSense/HardwareIO/data/config/nugus1/HardwareIO.yaml
Comment (jpptm) [reply module/platform/NUSense/HardwareIO/data/config/nugus1/HardwareIO.yaml:1]: I think @ysims can explain this better than me but I think we put each robot specific config in hwio because as far as I know, we do alias matching stuff with the robot configs now so that we don't have to replace config values every time a NUC is taken out or replaced

Comment (Bidski) [reply module/platform/NUSense/HardwareIO/data/config/nugus1/HardwareIO.yaml:1]: The default config should have all values set to sane defaults. The robot-specific configs should only set the values that differ from the default config.

File: module/platform/NUSense/HardwareIO/src/NUSenseParser.hpp (hunk 24 above)
Comment (TrentHouliston) [reply module/platform/NUSense/HardwareIO/src/NUSenseParser.hpp:12]: ///< is a totally viable alternate doxygen syntax when you want to be more concise!

File: module/platform/NUSense/HardwareIO/data/config/nugus1/HardwareIO.yaml
Comment (TrentHouliston) [reply module/platform/NUSense/HardwareIO/data/config/nugus1/HardwareIO.yaml:1]: I would also say name this file NUSense.yaml as it's going to make your life easier when trying to build for multiple robots

File: module/tools/FirmwareInstaller/src/FirmwareInstaller.cpp (hunk 26 above)
Comment (TrentHouliston) [reply module/tools/FirmwareInstaller/src/FirmwareInstaller.cpp:60]: Also that just seems like a bug

Comment (jpptm) [reply module/tools/FirmwareInstaller/src/FirmwareInstaller.cpp:60]: Cant believe this one got past the build checks lol, should get fixed by tomorrow hopefully

File: module/platform/NUSense/HardwareIO/src/NUSenseParser.hpp (hunk 24 above)
Comment (Bidski) [reply module/platform/NUSense/HardwareIO/src/NUSenseParser.hpp:12]: Consistency trumps "valid alternative style"

File: module/tools/FirmwareInstaller/src/FirmwareInstaller.cpp (hunk 26 above)
Comment (ysims) [reply module/tools/FirmwareInstaller/src/FirmwareInstaller.cpp:60]: This is because this module isn't in any roles, so it's not being built.

File: module/platform/NUSense/HardwareIO/data/config/nugus1/HardwareIO.yaml
Comment (ysims) [reply module/platform/NUSense/HardwareIO/data/config/nugus1/HardwareIO.yaml:1]: What @bidski is saying has nothing to do with the robot name config type. 
Configs are merged from default to more specific, so if a field is always the same for all robots, ie everything that isn't under `servos` here, can be put in default and not included here.

I'm not sure if renaming to NUSense.yaml will do much for building to robots. But it would mean we wouldn't need to restrict to only one HardwareIO module being built and can have roles for each explicitly for CI.  The other thing is that these and the OpenCR HardwareIOs will need to be the same, and updating the same thing in two places is annoying.

File: module/platform/NUSense/HardwareIO/src/NUSenseParser.hpp (hunk 24 above)
Comment (ysims) [reply module/platform/NUSense/HardwareIO/src/NUSenseParser.hpp:12]: I kinda like it for short things

File: module/input/FakeCamera/src/FakeCamera.cpp (hunk 21 above)
Comment (JohanneMontano) [reply module/input/FakeCamera/src/FakeCamera.cpp:85]: StreamReactor uses `dedent` which relies on other functions in strutil like `join` and the like. I will try to be consistent but I think I will have to definitely add stuff into strutil and make it a part of this pr otherwise it wont compile

File: module/platform/NUSense/HardwareIO/data/config/nugus1/HardwareIO.yaml
Comment (JohanneMontano) [reply module/platform/NUSense/HardwareIO/data/config/nugus1/HardwareIO.yaml:1]: My bad I just didnt understand it at all, not enough coffee

File: module/platform/NUSense/HardwareIO/src/NUgus.cpp (hunk 25 above)
Comment (JohanneMontano) [reply module/platform/NUSense/HardwareIO/src/NUgus.cpp]: Removed.

File: module/platform/NUSense/HardwareIO/src/NUSenseParser.hpp (hunk 24 above)
Comment (JohanneMontano) [reply module/platform/NUSense/HardwareIO/src/NUSenseParser.hpp]: I chose consistency

File: module/tools/FirmwareInstaller/src/FirmwareInstaller.cpp (hunk 26 above)
Comment (JohanneMontano) [reply module/tools/FirmwareInstaller/src/FirmwareInstaller.cpp]: Reverted.

File: module/platform/NUSense/HardwareIO/data/config/nugus1/HardwareIO.yaml
Comment (JohanneMontano) [reply module/platform/NUSense/HardwareIO/data/config/nugus1/HardwareIO.yaml:1]: Addressed.

File: module/platform/NUSense/HardwareIO/src/NUSenseParser.cpp [hunk 27]
```diff
@@ -1,11 +1,18 @@
 #include "NUSenseParser.hpp"
 
 namespace module::platform::NUSense {
-
+    /// @brief  Read a buffer of bytes and convert it to an unsigned 32 bit integer in little endian
+    ///         HwIO uses this to read the size field of the nbs packet sent by NUSense
+    /// @param  ptr Pointer to a buffer of bytes
+    /// @return The unsigned 32 bit number decoded from the buffer of bytes
```
Comment (Bidski) [inline module/platform/NUSense/HardwareIO/src/NUSenseParser.cpp:7]:
```suggestion
    /**
     * @brief  Read a buffer of bytes and convert it to an unsigned 32 bit integer in little endian
     *         HwIO uses this to read the size field of the nbs packet sent by NUSense
     * @param  ptr Pointer to a buffer of bytes
     * @return The unsigned 32 bit number decoded from the buffer of bytes
     */
```

File: shared/utility/strutil/strutil.hpp
Comment (Bidski) [inline shared/utility/strutil/strutil.hpp:1]: Why are these changes here?

**Review by Bidski — APPROVED** [review]

File: shared/utility/strutil/strutil.hpp
Comment (JohanneMontano) [reply shared/utility/strutil/strutil.hpp:1]: Its because StreamReactor uses dedent which was not in this file in main, but dedent relied on other functions such as join so I had to add those in as well

**Review by ysims — APPROVED** [review]
